Proposal: Extract ActionRunner interface

nimrodshn opened the pull request:
## What this does?

- Extracting run*Job to a ~~JobRunner~~ ActionRunner interface and adds structs to implement this interface.
- Adds a ~~JobRunner~~ ActionRunner map to healer.

## Why do this?

1. This extracts away methods and fields from healer - an already very heavy struct and creates ~~JobRunners~~ ActionRunner specific method and implementations which healer shouldn't be aware of.
2. This also, allows to extend ~~JobRunners~~ ActionRunner to any number and variety (If someday we would like to add more options more targets for jobs)
3. And most importantly; it allows me to stub for `RunJob` and by that write tests for `alerts_worker` and other parts of healer which currently not testable :cry: .

cc: @yaacov @moolitayer @jhernand

yaacov commented on cmd/autoheal/alerts_worker.go:
```diff
@@ -219,9 +219,10 @@ func (h *Healer) runRule(rule *autoheal.HealingRule, alert *alertmanager.Alert)
 	// Execute the action:
 	switch typed := action.(type) {
 	case *autoheal.AWXJobAction:
```
can we have the rule have a "runner" field, so we can just do `rule.run()` ?
I think it will be much more fun then this case ...

nimrodshn replied:
Sure, good suggestion!

jhernand replied:
I'd prefer to avoid that and keep the rule object as a pure container of data, without logic.

nimrodshn replied:
I tend to agree with @jhernand

jhernand commented on cmd/autoheal/active_jobs_worker.go:
```diff
@@ -21,15 +21,15 @@ import (
 )
 
-func (h *Healer) runActiveJobsWorker() {
+func (r AWXJobRunner) runActiveJobsWorker() {
```
`r` should be a pointer, otherwise you will be copying the contents of the struct each time you call this method (and any other method that has the struct as receiver):

```go
func (r *AWXJobRunner) runActiveJobsWorker() {
  ...
}
```

jhernand commented on cmd/autoheal/alerts_worker.go:
```diff
@@ -219,9 +219,10 @@ func (h *Healer) runRule(rule *autoheal.HealingRule, alert *alertmanager.Alert)
+
 	case *batch.Job:
-		err = h.runBatchJob(rule, typed, alert)
+		err = h.jobRunners[BATCH].RunJob(rule, typed, alert)
```
What is the benefit of this map if we always index it with constants? Wouldn't it be simpler to have two fields?

```go
type Healer struct {
  ...
  JobRunner awxRunner;
  JobRunner batchRunner;
  ...
}

nimrodshn replied:
I just thought it would be easier to scale with a map if we eventually in the future add Runners..

jhernand commented on cmd/autoheal/awx_job.go:
```diff
@@ -162,3 +171,7 @@ func (h *Healer) checkAWXJobStatus(jobID int) (finished bool, err error) {
 	return
 }
+
+func (r AWXJobRunner) AWX() *config.AWXConfig {
```
Using this method makes calls shorter, but also less clear: it is no longer explicit that you are using the configuration. I'd suggest to avoid that.

jhernand commented on cmd/autoheal/healer.go:
```diff
@@ -147,8 +147,17 @@ func (b *HealerBuilder) Build() (h *Healer, err error) {
 
-	// initialize active jobs map:
-	h.activeJobs = new(syncmap.Map)
+	h.jobRunners = map[JobRunnerType]JobRunner{
```
If we are using an interface to hide the implementation details of each job runner, then we should also hide the implementation details of the creation of the job runner, we should not access directly internal fields like `config` or `activeJobs`. To enforce this I'd suggest that you move the implementations of the job runners to different packages, and make everything private, unless really needed. That will force you to create a clean interface between the healer and the job runner.

nimrodshn replied:
Working on it :+1:

jhernand commented on cmd/autoheal/job_runner.go:
```diff
@@ -0,0 +1,33 @@
+type JobRunnerType int
+
+const (
+	AWX   JobRunnerType = 0
```
I'd prefer to avoid this kind of constants, as the fact that there are different types of job runners is already indicated by the existence of the corresponding Go types. If the constants are needed then please consider using `reflect.TypeOf` or `reflect.TypeOf(...).Name()` instead.

Also, for constants try to use camel case, and add a prefix that corresponds to the type name:

```go
const (
   JobRunnerTypeAWX JobRunnerType = iota
   JobRunnerTypeBatch
)
```

jhernand commented on cmd/autoheal/job_runner.go:
```diff
@@ -0,0 +1,33 @@
+	BATCH JobRunnerType = 1
+)
+
+type JobRunner interface {
```
I am not 100% sure if `JobRunner` is the right name for this interface. Currently it is running either an AWX job or a Kubernetes batch Job, thus the _job_ prefix makes sense. But in general what it runs is a _healing action_, which may be other thing. For example, if in the future we add support for Ansible Runner then it will be running an Ansible Playbook, not a _job_. So consider renaming it to `HealingActionRunner`, or something similar.

nimrodshn replied:
Sounds good :+1:

moolitayer replied:
@nimrodshn I gave the same comment when we talked about this but did't explain it as well...

moolitayer commented on cmd/autoheal/healer.go:
```diff
@@ -147,8 +147,17 @@ func (b *HealerBuilder) Build() (h *Healer, err error) {
-	h.activeJobs = new(syncmap.Map)
+	h.jobRunners = map[JobRunnerType]JobRunner{
+		AWX: AWXJobRunner{
+			config: h.config.AWX(),
```
When constructing the implementations It would be clean if they each get the entire configuration, and select the pieces relevant to them.

nimrodshn replied:
Nice thanks @moolitayer

moolitayer commented on cmd/autoheal/healer.go:
```diff
@@ -147,8 +147,17 @@ func (b *HealerBuilder) Build() (h *Healer, err error) {
+		AWX: AWXJobRunner{
+			config: h.config.AWX(),
+			// initialize active jobs map:
+			activeJobs: new(syncmap.Map),
```
This does not have to be part this PR, but it would be cleaner if the creator of an AWXJobRunner would not have to supply this map. It could be done in a constructor.

nimrodshn replied:
Yeah thats a good comment I just noticed that going over the code again :+1:

nimrodshn commented:
@moolitayer @jhernand @yaacov  Addressing you're comments I have done the following:

1. Extracted the `JobRunners` to `pkg/job_runners`
2. Extracted `metrics_exporter.go` from healer to a separate `pkg` so the runners can use it.
3. Fixed existing tests to use this `Runner` pattern.
4. Tested that healer works well with this new pattern.

Things to do in following PR's moving forward:
1. Move `Init` method *outside* the structs implementing `Runner` and into the `Runner` itself thus having a clean interface.
2. Other ideas are welcome here to enhance this patters :grinning: .

yaacov commented:
LGTM :+1:

nimrodshn commented:
@jhernand @moolitayer mind taking a look?

jhernand commented on pkg/job_runner/awx_job_runner/awx_job_runner.go:
```diff
@@ -14,34 +14,57 @@ See the License for the specific language governing permissions and
+	return awxJubRunner
+}
 
+func (r *AWXJobRunner) Init(config *config.Config, stopCh <-chan struct{}) *AWXJobRunner {
```
This kind of constructor in two steps is problematic, in my opinion, specially the fact that the `Init` method receives multiple parameters. What if we need or want to add a new optional parameter in the future? We will have to change all the points where this method is called. That isn't very relevant in this case, because it is used only internally, but in general I prefer to use a builder:

```go
type Builder {
  ...
}

func NewBuilder() *Builder {
  return new(Builder)
}

// Config sets the configuration that will be used by the
// runner.
func (b *Builder) Config(config *config.Config) *Builder {
  b.config = config
  return b
}

// StopChannel sets the channel where the runner will wait
// for the signal to stop its background processing.
func (b *Builder) StopChannel(ch <-chan struct{}) *Builder {
  b.stopCh = ch
  return b
}

// Build creates a new runner using the details stored in
// the builder.
func (b *Builder) Build() (runner *AWXJobRunner, err error) {
  // Create and populate the runner, and also check for
  // possible errors, like missing mandatory parameters,
  // etc.
  runner = ...

  return
}
```

This way the user of this package can create the runner like this:

```go
runner, err := awxrunner.NewBuilder().
  Config(config).
  StopChannel(stopCh).
  Build()
if err != nil {
  ...
}
```

If we eventually decide to add optional parameters to that constructor, that code doesn't need to change.

Can you use this mechanism for both kinds of runners?

nimrodshn replied:
Yeah very nice! Actoually I never really agreed with this approach in the code, but now i'm sold :+1: :trophy:

jhernand commented on pkg/job_runner/awx_job_runner/awx_job_runner.go:
```diff
@@ -14,34 +14,57 @@ See the License for the specific language governing permissions and
-	awxPassword := h.config.AWX().Password()
-	awxCA := h.config.AWX().CA()
-	awxInsecure := h.config.AWX().Insecure()
+	awxAddress := r.AWX().Address()
```
I think that this `AWX` method unnecessary hides the fact that we are accessing the configuration. Please avoid it.

jhernand commented on pkg/job_runner/awx_job_runner/awx_job_runner.go:
```diff
@@ -167,3 +190,7 @@ func (h *Healer) checkAWXJobStatus(jobID int) (finished bool, err error) {
 	return
 }
+
+func (r *AWXJobRunner) AWX() *config.AWXConfig {
```
I think this method isn't necessary.

jhernand commented on cmd/autoheal/healer.go:
```diff
@@ -162,7 +166,16 @@ func (h *Healer) Run(stopCh <-chan struct{}) error {
+	h.actionRunners[ActionRunnerTypeAWX] = awxrunner.NewAWXActionRunnerBuilder().
+		Config(h.config.AWX()).
+		StopCh(stopCh).
+		Build()
```
I think we should check the errors returned by the `Build` methods.

jhernand commented on pkg/awxrunner/awx_action_runner.go:
```diff
@@ -14,34 +14,73 @@ See the License for the specific language governing permissions and
 
-func (h *Healer) runAWXJob(rule *autoheal.HealingRule, action *autoheal.AWXJobAction, alert *alertmanager.Alert) error {
-	var err error
+type AWXActionRunnerBuilder struct {
```
As this has its own package now I think you can rename it to just `Builder`. Same for the runner, it can be just `Runner`.

jhernand commented on pkg/awxrunner/awx_action_runner.go:
```diff
@@ -167,3 +206,7 @@ func (h *Healer) checkAWXJobStatus(jobID int) (finished bool, err error) {
 	return
 }
+
+func (r *AWXActionRunner) AWX() *config.AWXConfig {
```
I still think that this method isn't needed, and using it hides the fact that we are using the configuration. Can you remove this method and use directly `config...`?

nimrodshn replied:
Yeah I forgot to remove it :cry:

jhernand commented on pkg/batchrunner/batch_action_runner.go:
```diff
@@ -14,28 +14,60 @@ See the License for the specific language governing permissions and
+	return b
+}
+
+func (b *BatchActionRunnerBuilder) Build() *BatchActionRunner {
```
The build method should return the built object an and an error. We may not need the error today, but it will most likely needed in the future. Please add it.

jhernand commented on pkg/awxrunner/awx_action_runner.go:
```diff
@@ -14,34 +14,73 @@ See the License for the specific language governing permissions and
+	return b
+}
 
+func (b *AWXActionRunnerBuilder) Build() *AWXActionRunner {
```
Build should return the result and an error.

jhernand commented on pkg/metrics/metrics_exporter.go:
```diff
@@ -24,15 +24,19 @@ var (
 	)
 )
 
-func (h *Healer) metricsHandler() http.Handler {
```
As this has its own package now, you can simplify the names, just `Handler`, for example. That way the user of this package will write `metrics.Handler` which reads nicely.

nimrodshn replied:
nice :+1:

nimrodshn commented:
@jhernand Please take a look :+1:

jhernand commented on cmd/autoheal/healer.go:
```diff
@@ -162,7 +166,24 @@ func (h *Healer) Run(stopCh <-chan struct{}) error {
+	awxRunner, err := awxrunner.NewBuilder().
+		Config(h.config.AWX()).
+		StopCh(stopCh).
+		Build()
```
We need to check `err` here as well, otherwise if this fails but the next does't then we don't detect it.

nimrodshn commented:
@jhernand Done.
